csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs

```diff
@@ -59,6 +97,75 @@ protected async ValueTask<RecordBatch> ReadRecordBatchAsync(CancellationToken ca
         {
             await ReadSchemaAsync().ConfigureAwait(false);
 
+            await ReadInitialDictionariesAsync(cancellationToken).ConfigureAwait(false);
+
+            return await ReadArrowObjectAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+
+        protected RecordBatch ReadRecordBatch()
+        {
+            ReadSchema();
+
+            ReadInitialDictionaries();
+
+            return ReadArrowObject();
```

Comment:
The [spec](https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format) says

>  DictionaryBatch and RecordBatch messages may be interleaved

This appears to only support DictionaryBatches above RecordBatches. If we read all the "initial" dictionaries, and then read a RecordBatch, the next time `ReadRecordBatch()` is called, if there is a DictionaryBatch, `null` will be returned. Is that intentional?

Reply:
I tested reading interleaved dictionaries by creating a test file with python and reading the test file with C#, and it seemed to work fine.
The code for creating the test file is below, this code is based on the test code of the python implementation.

https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_ipc.py#L417

```
import pyarrow as pa

ty = pa.dictionary(pa.int8(), pa.utf8())
data = [["foo", "foo", None],
            ["foo", "bar", "foo"],
            ["foo", "bar"],
            ["foo", None, "bar", "quux"], 
            ["bar", "quux"],
            ]
batches = [
        pa.RecordBatch.from_arrays([pa.array(v, type=ty)], names=['dicts'])
        for v in data]
schema = batches[0].schema

def write_batches():
    with pa.RecordBatchStreamWriter("./dictionary_batch_test.batch", schema = schema) as writer:
        for batch in batches:
            writer.write_batch(batch)

st = write_batches()
```

We should add C# tests for this after `ArrowStreamWriter` supports writing interleaved dictionaries.

Reply:
> I would like to implement it in another PR because this PR will become too large.
How is that?

That is perfectly reasonable to me. I think the current functionality is sufficient for the first PR.

Reply:
> I tested reading interleaved dictionaries by creating a test file with python

In my spare time, I've been trying to set up the integration tests, so we can test the C# implementation against the other languages. Hopefully I can get the initial PR for that up soon.